model/test_results_test.go

```diff
@@ -0,0 +1,468 @@
+	db := env.GetDB()
+	ctx, cancel := env.Context()
+	defer cancel()
+	defer func() {
+		assert.NoError(t, db.Collection(testResultsCollection).Drop(ctx))
+	}()
+
+	tr1 := getTestResults()
+	_, err := db.Collection(testResultsCollection).InsertOne(ctx, tr1)
+	require.NoError(t, err)
+	tr2 := getTestResults()
+	_, err = db.Collection(testResultsCollection).InsertOne(ctx, tr2)
+	require.NoError(t, err)
+
+	t.Run("NoEnv", func(t *testing.T) {
+		tr := &TestResults{ID: tr1.ID, populated: true}
+		assert.Error(t, tr.Close(ctx))
+	})
+	t.Run("DNE", func(t *testing.T) {
+		tr := &TestResults{ID: "DNE"}
+		tr.Setup(env)
+		assert.Error(t, tr.Close(ctx))
+	})
+	t.Run("WithID", func(t *testing.T) {
+		tr := &TestResults{ID: tr1.ID, populated: true}
+		tr.Setup(env)
+		require.NoError(t, tr.Close(ctx))
+
+		updated := &TestResults{}
+		require.NoError(t, db.Collection(testResultsCollection).FindOne(ctx, bson.M{"_id": tr1.ID}).Decode(updated))
+		assert.Equal(t, tr1.ID, updated.ID)
+		assert.Equal(t, tr1.ID, updated.Info.ID())
+		assert.Equal(t, tr1.CreatedAt, updated.CreatedAt)
+		assert.True(t, time.Since(updated.CompletedAt) <= time.Second)
+		assert.Equal(t, tr1.Info, updated.Info)
+		assert.Equal(t, tr1.Artifact, updated.Artifact)
+	})
+	t.Run("WithoutID", func(t *testing.T) {
+		tr := &TestResults{Info: tr2.Info, populated: true}
+		tr.Setup(env)
+		require.NoError(t, tr.Close(ctx))
+
+		updated := &TestResults{}
+		require.NoError(t, db.Collection(testResultsCollection).FindOne(ctx, bson.M{"_id": tr1.ID}).Decode(updated))
+		assert.Equal(t, tr1.ID, updated.ID)
+		assert.Equal(t, tr1.ID, updated.Info.ID())
+		assert.Equal(t, tr1.CreatedAt, updated.CreatedAt)
+		assert.True(t, time.Since(updated.CompletedAt) <= time.Second)
+		assert.Equal(t, tr1.Info, updated.Info)
+		assert.Equal(t, tr1.Artifact, updated.Artifact)
+	})
+}
+
+func getTestResults() *TestResults {
+	info := TestResultsInfo{
+		Project:   utility.RandomString(),
+		Version:   utility.RandomString(),
+		Variant:   utility.RandomString(),
+		TaskName:  utility.RandomString(),
+		TaskID:    utility.RandomString(),
+		Execution: seedRand.Intn(5),
+		Requester: utility.RandomString(),
+	}
+	return &TestResults{
+		ID:          info.ID(),
+		Info:        info,
+		CreatedAt:   time.Now().Add(-time.Hour).UTC().Round(time.Millisecond),
+		CompletedAt: time.Now().UTC().Round(time.Millisecond),
+		Artifact: TestResultsArtifactInfo{
+			Type:   PailLocal,
+			Prefix: info.ID(),
+		},
+	}
+}
+
+func getTestResult() TestResult {
+	return TestResult{
+		TestName:       utility.RandomString(),
+		Trial:          seedRand.Intn(10),
+		Status:         "Pass",
+		LineNum:        seedRand.Intn(1000),
+		TaskCreateTime: time.Now().Add(-time.Hour).UTC().Round(time.Millisecond),
+		TestStartTime:  time.Now().Add(-30 * time.Hour).UTC().Round(time.Millisecond),
+		TestEndTime:    time.Now().UTC().Round(time.Millisecond),
+	}
+}
+
+var seedRand *rand.Rand = rand.New(rand.NewSource(time.Now().UnixNano()))
```

Comment:
I'd just call rand.Seed() in init....  and call the rand package's init functions.